multi-tenancy.md

```diff
@@ -0,0 +1,55 @@
+
+* The resolver type/interface should have the following characteristics:
+    - It receives a HTTP request for inspection (some may also provide the response, if available).
+    - It returns an Organization if one can be resolved.
+    - It attaches the resolved Organization to the request, so that the developer can make use of this context.
+
+* The default organization resolver should have the following characteristics:
+
+    - If `stormpath.web.baseDomain` is defined, e.g. `example.com`, then a request with the `Host` header specified as `org-a.example.com` should return the `Organization` that has the `nameKey` of `org-a`.
+    - If the request body has a field of `organizationNameKey=org-a`, then the `Organization` which has the `nameKey` of `org-a` should be returned.  This covers the case where the user is supplying the org key in a form field.
+    - The subdomain value should take precedence over the form value.
+
+* If `stormpath.web.multiTenancy=true`, and the organization cannot be resolved, AND the developer has disabled the `organizationNameKey` for fields, then the operation (login, etc) will be posted against the application per normal.
+
+* When authenticating requests with access tokens, the `org` claim of of the access token must match the resolved organization for the request.  If this is
+not true the request should be rejected and the user should be required to authenticate for a new token.  If the access token was sent in a cookie header, the response should delete this cookie.
+
+* When processing a email verification request:
+    - If the key has expired or cannot be found, and an organization cannot be resolved, redirect the user to `<baseDomain>/verify`, and render a field that allows them to specify their organization name.
+    - If the key is valid, the REST API response needs to include the organization name key on the response, so that we can redirect the user to `<orgNameKey>.<baseDomain><verifyEmail.nextUri|login.nextUri>` - *REQUIRES REST API CHANGE*
+
+* When processing a password reset request:
+    - If the key has expired or cannot be found, and an organization cannot be resolved, redirect the user to `<baseDomain>/forgot`, and render a field that allows them to specify their organization name.
+    - If the key is valid, the REST API response needs to include the organization name key on the response, so that we can redirect the user to `<orgNameKey>.<baseDomain><changePassword.nextUri|login.nextUri>` - *REQUIRES REST API CHANGE*
+
+* When the user submits an organization name key, and that organization does not exist, the error message to the end user should not leak information about organizations that do or do not exist.  In this situation the error message should read to the effect "Username or password is invalid, or Organization does not exist".
+
+## Future Use Cases
+
+Here are some notes we have collected about possible future use cases:
```

Comment:
My point is that it adds noise - it's not necessary in the actual spec document until its necessary.